spark2/src/main/java/org/apache/iceberg/spark/source/Reader.java

```diff
@@ -166,13 +167,17 @@
     this.readTimestampWithoutZone = SparkUtil.canHandleTimestampWithoutZone(options.asMap(), sessionConf);
   }
 
+  protected Schema snapshotSchema() {
+    return SnapshotUtil.schemaFor(table, snapshotId, asOfTimestamp);
```

Comment:
I don't think that this should determine the snapshot that will be scanned in multiple places, but right now the logic to find the snapshot is here and in `tasks()`.

Instead, I think this class should create a `TableScan`, configure it using the snapshot selection criteria, and use the `schema()` from the table scan here. In `tasks()`, this base scan can be further refined before getting tasks (each `TableScan` object is immutable and independent).

Reply:
I have a working version of this. I'll post a PR against your branch.

Reply:
https://github.com/wypoon/iceberg/pull/1

Reply:
Incorporated refactored version of wypoon#1.